go/test/endtoend/shardedrecovery/sharded_recovery_test.go

```diff
@@ -0,0 +1,630 @@
+	if err != nil {
+		time.Sleep(10 * time.Millisecond)
+		checkData(t, tablet, count)
+	}
+	if len(qr.Rows) > 0 {
+		countResult, err := strconv.Atoi(fmt.Sprintf("%s", qr.Rows[0][0].ToBytes()))
+		require.Nil(t, err)
+		if countResult != count {
+			checkData(t, tablet, count)
+		}
+	}
+}
+
+func removeTablets(t *testing.T, tablets []*cluster.Vttablet) {
+	var mysqlProcs []*exec.Cmd
+	for _, tablet := range tablets {
+		proc, _ := tablet.MysqlctlProcess.StopProcess()
+		mysqlProcs = append(mysqlProcs, proc)
+		tablet.VttabletProcess.TearDown()
+	}
+	for _, proc := range mysqlProcs {
+		err := proc.Wait()
+		assert.Nil(t, err)
+	}
+}
+
+func restoreTablet(t *testing.T, tablet *cluster.Vttablet, restoreKSName string, shardName string) {
+	err := cluster.ResetTabletDirectory(*tablet)
+	assert.Nil(t, err)
+	tm := time.Now().UTC()
+	tm.Format(time.RFC3339)
+	_, err = localCluster.VtctlProcess.ExecuteCommandWithOutput("CreateKeyspace",
+		"-keyspace_type=SNAPSHOT", "-base_keyspace="+keyspaceName,
+		"-snapshot_time", tm.Format(time.RFC3339), restoreKSName)
+	assert.Nil(t, err)
+
+	replicaTabletArgs := commonTabletArg
+	replicaTabletArgs = append(replicaTabletArgs, "-disable_active_reparents",
+		"-enable_replication_reporter=false",
+		"-init_tablet_type", "replica",
+		"-init_keyspace", restoreKSName,
+		"-init_shard", shardName)
+	tablet.VttabletProcess.SupportsBackup = true
+	tablet.VttabletProcess.ExtraArgs = replicaTabletArgs
+
+	tablet.VttabletProcess.ServingStatus = ""
+	err = tablet.VttabletProcess.Setup()
+	assert.Nil(t, err)
+
+	err = tablet.VttabletProcess.WaitForTabletTypesForTimeout([]string{"SERVING"}, 20*time.Second)
+	assert.Nil(t, err)
+}
+
+func verifyQueriesUsingVtgate(t *testing.T, session *vtgateconn.VTGateSession, query string, value string) {
+	qr, err := session.Execute(context.Background(), query, nil)
+	assert.Nil(t, err)
+	assert.Equal(t, value, fmt.Sprintf("%v", qr.Rows[0][0]))
+}
+
+func executeQueriesUsingVtgate(t *testing.T, session *vtgateconn.VTGateSession, query string) {
+	_, err := session.Execute(context.Background(), query, nil)
+	assert.Nil(t, err)
+}
+
+func initializeCluster() (int, error) {
+
+	localCluster = cluster.NewCluster(cell, hostname)
+
+	// Start topo server
+	err := localCluster.StartTopo()
+	if err != nil {
+		return 1, err
+	}
+	// Start keyspace
+	keyspace := &cluster.Keyspace{
+		Name: keyspaceName,
+	}
+	localCluster.Keyspaces = append(localCluster.Keyspaces, *keyspace)
+
+	shard := cluster.Shard{
+		Name: shardName,
+	}
+	shard0 := cluster.Shard{
+		Name: shard0Name,
+	}
+	shard1 := cluster.Shard{
+		Name: shard1Name,
+	}
+
+	var mysqlProcs []*exec.Cmd
+	for _, eachShard := range []cluster.Shard{shard, shard0, shard1} {
+		tabletsRequired := 3
+		if eachShard.Name == shardName {
+			tabletsRequired = 5
+		}
+		for i := 0; i < tabletsRequired; i++ {
+			tabletType := "replica"
+			if i == 0 {
+				tabletType = "master"
```

Comment:
You don't need to do this if you are doing initShard master

Reply:
done